pilot/pkg/networking/core/envoyfilter/rc_patch.go

```diff
@@ -271,23 +271,23 @@ func routeConfigurationMatch(patchContext networking.EnvoyFilter_PatchContext, r
 	// we match on the port number and virtual host for sidecars
 	// we match on port number, server port name, gateway name, plus virtual host for gateways
 	if patchContext != networking.EnvoyFilter_GATEWAY {
-		listenerPort := 0
-		if strings.HasPrefix(rc.Name, string(model.TrafficDirectionInbound)) {
-			_, _, _, listenerPort = model.ParseSubsetKey(rc.Name)
-		} else {
-			listenerPort, _ = strconv.Atoi(rc.Name)
+		if rMatch.Name != "" && rMatch.Name != rc.Name {
+			return false
 		}
-
 		// FIXME: Ports on a route can be 0. the API only takes uint32 for ports
 		// We should either make that field in API as a wrapper type or switch to int
-		if rMatch.PortNumber != 0 && int(rMatch.PortNumber) != listenerPort {
-			return false
-		}
+		if rMatch.PortNumber != 0 {
```

Comment:
nit: would it also make sense to abstract away some of the listenerPort computation out of the function to make it a little bit more readable too? That way instead of having 3 levels of if statements with extra logic for getting the listener port from a route configuration name..

e.g.:
```go
if patchContext != networking.EnvoyFilter_GATEWAY {
	if rMatch.Name != "" && rMatch.Name != rc.Name {
		return false
	}
	// FIXME: Ports on a route can be 0. the API only takes uint32 for ports
	// We should either make that field in API as a wrapper type or switch to int
	if rMatch.PortNumber != 0 {
		listenerPort := 0
		if strings.HasPrefix(rc.Name, string(model.TrafficDirectionInbound)) {
			_, _, _, listenerPort = model.ParseSubsetKey(rc.Name)
		} else {
			listenerPort, _ = strconv.Atoi(rc.Name)
		}
```

, we can just have 2 levels of if statements.. and `routeConfigurationMatch()` can focus primarily on route config matching with listenerPort computation happening in it's own function.

```go
func routeConfigurationMatch(patchContext networking.EnvoyFilter_PatchContext, rc *route.RouteConfiguration,
	rp *model.EnvoyFilterConfigPatchWrapper, portMap model.GatewayPortMap,
) bool {
	rMatch := rp.Match.GetRouteConfiguration()
	if rMatch == nil {
		return true
	}

	// we match on the port number and virtual host for sidecars
	// we match on port number, server port name, gateway name, plus virtual host for gateways
	if patchContext != networking.EnvoyFilter_GATEWAY {
		if rMatch.Name != "" && rMatch.Name != rc.Name {
			return false
		}
		// FIXME: Ports on a route can be 0. the API only takes uint32 for ports
		// We should either make that field in API as a wrapper type or switch to int
		listenerPort := getListenerPortFromName(rc.Name)
		if rMatch.PortNumber != 0 && int(rMatch.PortNumber) != listenerPort {
			return false
		}
		return true
	}

	// This is a gateway. Get all the fields in the gateway's RDS route name
	routePortNumber, portName, gateway := model.ParseGatewayRDSRouteName(rc.Name)
	if rMatch.PortNumber != 0 && !anyPortMatches(portMap, routePortNumber, int(rMatch.PortNumber)) {
		return false
	}
	if rMatch.PortName != "" && rMatch.PortName != portName {
		return false
	}
	if rMatch.Gateway != "" && rMatch.Gateway != gateway {
		return false
	}

	if rMatch.Name != "" && rMatch.Name != rc.Name {
		return false
	}

	return true
}

func getListenerPortFromName(name string) int {
	listenerPort := 0
	if strings.HasPrefix(name, string(model.TrafficDirectionInbound)) {
		_, _, _, listenerPort = model.ParseSubsetKey(name)
	} else {
		listenerPort, _ = strconv.Atoi(name)
	}
	return listenerPort
}
```